### PR TITLE
[FeG][SessionD] Add access timezone information to requests to policy component

### LINKED ISSUE #2668: [FeG] Add in timezone AVP into CCRs going from FeG->PCRF/OCS

### PR DESCRIPTION
**Pull request by themarwhal**
Signed-off-by: Marie Bremner <marwhal@fb.com>

## Summary
This PR adds the access timezone AVP for CCR-Is. 
Document for AVP:
https://www.developingsolutions.com/DiaDict/Dictionary/3GPP-MS-TimeZone.html
Document for Timezone encoding: 3GPP TS 23.040, Section 9.2.3.11 TP-Service-Centre-Time-Stamp

SessionD loads the timezone information on SessionD start, and stores it in LocalEnforcer. This value is used for all session creations until SessionD restarts.

I've added this AccessTimezone field as optional field configured by `sessiond.yml`. This is to prevent regression in existing CWF deployments. (I'm still investigating on reliably configuring the SessionD docker container to have the host's timezone for CWF.) Also it will be useful for testing to be able to toggle the feature. The config name is `send_access_timezone`.

## Test Plan
Some basic unit tests for the encoding
Regression testing for FeG, SessionManager, CWF
Running manual tests with manually changing timezones on VM
1. Manually modify SessionD's timezone inside the docker container. I did this by modifying the docker-compose
```
  sessiond:
    volumes:
      - "/etc/timezone:/etc/timezone:ro"
      - "/etc/localtime:/etc/localtime:ro"
```
2. Run Integ test with GRPC logging CCR logging on SessionProxy
2a. Without `send_access_timezone` flag in SessionD.yml (Default for CWF)
```
I0918 19:50:58.036955       1 logcodec.go:46] Received: *protos.CreateSessionRequest: {
	"access_timezone": null,
	"session_id": "IMSI120230650810660-533732",
         ....
}
```
And no timezone AVP in Gx/Gy CCR-I
2b. Run with `send_access_timezone` flag 
```
I0918 20:17:32.177237       1 logcodec.go:46] Received: *protos.CreateSessionRequest: {
	"access_timezone": {
		"offset_minutes": -480
	},
	"session_id": "IMSI717809013284033-804419",
        .....
}
```
and timezone AVP for Gx CCR-I
```
I0918 20:35:17.961195       1 gx_client.go:132] Sending Gx CCR message
Credit-Control-Request (CCR)
{Code:272,Flags:0xc0,Version:0x1,Length:432,ApplicationId:16777238,HopByHopId:0x2aef752c,EndToEndId:0xb6c615e6}
	Session-Id {Code:263,Flags:0x40,Length:64,VendorId:0,Value:UTF8String{feg.magma.svc.cluster.local;144;854;IMSI594008207667977},Padding:1}
...
	TGPP-MS-TimeZone {Code:23,Flags:0x80,Length:16,VendorId:10415,Value:OctetString{0x2b},Padding:3}
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: feg/gateway/services/session_proxy/credit_control/gy/gy_client.go [hunk 1]
```diff
@@ -300,8 +300,6 @@ func getServiceInfoAvp(server *diameter.DiameterServerConfig, request *CreditCon
 	psInfoAvps := []*diam.AVP{
 		// Set PDP Type as IPV4(0)
 		diam.NewAVP(avp.TGPPPDPType, avp.Vbit, diameter.Vendor3GPP, datatype.Enumerated(0)),
```
Comment (themarwhal) [inline feg/gateway/services/session_proxy/credit_control/gy/gy_client.go:302]: Don't need it for Gy apparently https://www.developingsolutions.com/DiaDict/Dictionary/3GPP-MS-TimeZone.html

File: feg/gateway/services/session_proxy/credit_control/gx/gx_client.go [hunk 2]
```diff
@@ -486,3 +489,24 @@ func getDefaultFramedIpv4Addr() net.IP {
 	}
 	return ipV4V6
 }
+
+// TS 23.040 Section 9.2.3.11
```
Comment (uri200) [inline feg/gateway/services/session_proxy/credit_control/gx/gx_client.go:493]: 👍

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 3]
```diff
@@ -2004,6 +2004,32 @@ void LocalEnforcer::remove_rule_due_to_bearer_creation_failure(
...
+  }
+  int hours            = std::stoi(s.substr(0, 3), nullptr, 10);
+  int minutes          = std::stoi(s[0] + s.substr(3), nullptr, 10);
+  int total_offset_min = hours * 60 + minutes;
+  MLOG(MINFO) << "Access timezone is UTC " << (total_offset_min >= 0 ? "+" : "")
+              << total_offset_min << " minutes";
+  timezone.set_offset_minutes(total_offset_min);
+  return std::make_unique<Timezone>(timezone);
+}
+
```
Comment (uri200) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp:2032]: just curious, but why don't you just send the UTC/localtime time to FEG and leave FEG compute the timezone? This way the translation is all in one place and we may get to use that time for something else (you don't need to change that btw 😅)

**Review by uri200 — APPROVED** [review]

File: lte/gateway/c/session_manager/LocalEnforcer.cpp (hunk 3 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp:2032]: Oh yes, I suppose I could just do that too :D

**Review by ardzoht — APPROVED** [review]

**Pull request merged**